### PR TITLE
rustdoc: general cleanups

### PR DESCRIPTION
**Pull request by Mark-Simulacrum**
This is purely a refactoring, mostly just simplifying some of the code. Commits are best reviewed individually.

### CODE REVIEW

**Comment by JohnCSimon** [comment]
@QuietMisdreavus Howdy from triage! This PR has sat waiting for review for a week.

File: src/librustdoc/clean/mod.rs [hunk 1]
```diff
@@ -138,7 +137,10 @@ pub struct Crate {
     pub masked_crates: FxHashSet<CrateNum>,
 }
 
-impl<'a, 'tcx> Clean<Crate> for (visit_ast::RustdocVisitor<'a, 'tcx>, doctree::Module<'tcx>) {
+// The `()` here is rather ugly and would be great to remove. Unfortunately, we
+// already have a different Clean impl for `doctree::Module` which makes this
+// the only way to easily disambiguate.
+impl<'tcx> Clean<Crate> for ((), doctree::Module<'tcx>) {
```
Comment (eddyb) [inline src/librustdoc/clean/mod.rs]: Implement this on `&'tcx hir::Crate` instead and have it create `RustdocVisitor` here.

File: src/librustdoc/doctree.rs [hunk 2]
```diff
@@ -23,8 +23,7 @@ pub struct Module<'hir> {
     pub enums: Vec<Enum<'hir>>,
     pub fns: Vec<Function<'hir>>,
     pub mods: Vec<Module<'hir>>,
-    pub id: NodeId,
-    pub hid: hir::HirId,
+    pub id: hir::HirId,
```
Comment (eddyb) [inline src/librustdoc/doctree.rs]: Ideally, this entire module would use `DefId`, but this is an improvement.

**Comment by eddyb** [comment]
r? @eddyb 

r=me after rebase and maybe @petrochenkov can take a look at the resolver changes

**Comment by Mark-Simulacrum** [comment]
Rebased and fixed the review comments. I can try and replace everything with `DefId` in a follow up PR if you'd like; I believe this is waiting on @petrochenkov now for resolver change review.

**Review by petrochenkov — APPROVED** [review]

**Pull request merged**